<a href="https://colab.research.google.com/github/sebas123456m/pruebaColab/blob/main/Copy_of_Actividad_Python_basico.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1 style="text-align:center">
    Actividad de aprendizaje: Python básico
</h1>

<br>

<font color="red">

<h3 style="text-align:center">
    SEBASTIAN MENDEZ MONTEALEGRE
</h3>
    
<h3 style="text-align:center">
    Fecha 3 DE JUNIO DE 2023
    PROBANDO OTRA VEZ EN ESTE COMENTARIO PARA TRABAJAR CON GIT Y COLAB

    
    
</h3>

<hr>

## Introducción

<font color="darkblue"> Este cuaderno es una plantilla inicial de trabajo, realice una copia en la misma carpeta. Renombre la copia con su nombre, sin tildes ni eñes con el siguiente formato "**Apellido1_Apellido2_Nombre1_Nombre2.ipynb**" (Ejemplo: "**Ibanez_Forero_Luis_Eduardo.ipynb**") para trabajar sobre él.
 
- Cuando termine de resolver el proyecto, ejecutelo todo (**NO LE DEBE APARECER NINGUN ERROR**), despues descargue el proyecto en formato html, abralo y guadelo en formato pdf.
- Mis explicaciones van en azul y estas no deben ser modificadas, tampoco deben ser modificadas por ninguna razón las celdas donde diga "**Prueba de funcionamiento**". Su trabajo consiste en plantear las celdas necesarias, ya sean en formato *code* o formato *markdown* que le permita resolver el proyecto. 
</font>  

In [ ]:
################################################################################################################
################################################################################################################
# Espacio para las librerías que va a necesitar
# Ejemplo: import numpy as np
################################################################################################################
################################################################################################################









<font color="darkblue"> Ejercicio 1: Crear una variable llamada `a` que contenga el texto "**De aquel amor. De música ligera. Nada nos libra.**". Crear otra variable llamada `b` que contenga el texto "**Nada más queda**". Crear una tercera variable llamada `c` donde solo se tome el texto "**Nada más** " de `b`. Crear una cuarta variable llamada `d` donde repita el texto guardado en `b` cuatro veces. Por último, crear una variable llamada `e` que una todo el texto de la siguiente forma.
</font>   

"De aquel amor. De música ligera. Nada nos libra. Nada más queda. Nada más Nada más queda. Nada más queda. Nada más queda. Nada más queda. "




In [ ]:
################################################################################################################
## Este espacio debe ser llenado por el estudiante.
################################################################################################################

a="De aquel amor. De música ligera. Nada nos libra. "


b="Nada más queda  "
c="Nada más "
tamanoc= len(c)
print (tamanoc)
d=b*4
e=(a+b+c+d)
print(e)


9
De aquel amor. De música ligera. Nada nos libra. Nada más queda  Nada más Nada más queda  Nada más queda  Nada más queda  Nada más queda  


In [ ]:
################################################################################################################
## Prueba de funcionamiento
## Esta celda no debe ser modificada
################################################################################################################
import numpy as np
p11=isinstance(a, str) & (len(a)==49)
p12=isinstance(b, str) & (len(b)==16) 
p13=isinstance(c, str) & (len(c)==9)
p14=isinstance(d, str) & (len(d)==64)
p15=isinstance(e, str) & (len(e)==138)
#p1=p11 & p12 & p13 & p14 & p15
p1=p11 & p12 & p13 & p14 & p15
if p1==True:
  n1=1
else:
  n1=0

print(f"¡¡¡Felicidades!!! Tu nota hasta este momento es {n1} de 5" if p1==True else f"Hay un error, vuelve a revisar tu codigo. Tu nota hasta este momento es {n1} de 5" )
################################################################################################################
## Prueba de funcionamiento
## Esta celda no debe ser modificada
################################################################################################################

¡¡¡Felicidades!!! Tu nota hasta este momento es 1 de 5


<font color="darkblue"> 2) Crear un vector llamado $v=[789,796,\dots,922]$ que empiece en 789 y contenga los múltiplos de 7 usando la función `arange()` de la librería numpy. Después crear una matriz $A$ de tamaño $(4\times 5)$ usando la función `reshape()` de la librería numpy. Calcular la transpuesta de la matriz $A$ usando el comando $A.T$ y guardar el resultado en una variable llamada $B$. Por último, realizar el producto matricial de $A$ y $B$ usando el operador `@`.
</font>  


In [ ]:
from sys import version_info
################################################################################################################
## Este espacio debe ser llenado por el estudiante.
################################################################################################################

v = np.arange(789, 929, 7) # parametros numpy.arange([start, ]stop, [step, ], dtype=None)
"""
La función arange() devuelve un array NumPy que contiene los números en el rango especificado. Por ejemplo, np.arange(0, 5) devolverá el array [0, 1, 2, 3, 4], mientras que np.arange(1, 10, 2) devolverá el array [1, 3, 5, 7, 9].

La función arange() es muy útil para crear arrays de números en secuencia para usar en cálculos matemáticos y científicos.
"""
print(v)
print(len(v))
print("A")
print("*****")
A=np.reshape(v,(4,5))
"""
La función reshape() de la librería NumPy se utiliza para cambiar la forma (o dimensiones) de un array NumPy existente, sin modificar sus elementos.
La sintaxis básica de la función reshape() es la siguiente:
new_array = np.reshape(original_array, new_shape)
DONDE:
original_array es el array que deseas modificar
new_shape es una tupla que especifica la nueva forma que deseas para el array
EJEMPLO: new_array = np.reshape(original_array, (3, 4))
"""
print(A)
B= A.T
print("*******")
print("Transpuesta")
print(B)
C=A @ B
print("******")
print("El producto de la matriz es: ")  
print(C)

[789 796 803 810 817 824 831 838 845 852 859 866 873 880 887 894 901 908
 915 922]
20
A
*****
[[789 796 803 810 817]
 [824 831 838 845 852]
 [859 866 873 880 887]
 [894 901 908 915 922]]
*******
Transpuesta
[[789 824 859 894]
 [796 831 866 901]
 [803 838 873 908]
 [810 845 880 915]
 [817 852 887 922]]
******
El producto de la matriz es: 
[[3224535 3365060 3505585 3646110]
 [3365060 3511710 3658360 3805010]
 [3505585 3658360 3811135 3963910]
 [3646110 3805010 3963910 4122810]]


In [ ]:
################################################################################################################
## Prueba de funcionamiento
## Esta celda no debe ser modificada
################################################################################################################

p21=isinstance(v,np.ndarray) and v.shape==(20,)
p22=isinstance(A,np.ndarray) and A.shape==(4,5)
p23=isinstance(B,np.ndarray) and B.shape==(5,4)
p24=isinstance(C,np.ndarray) and C.shape==(4,4) and np.linalg.det(C)==0
#p2=p21 & p22 & p23 & p24 
p2=p21 & p22 & p23 & p24 
if p2==True:
  n2=1
else:
  n2=0

print(f"¡¡¡Felicidades!!! Tu nota hasta este momento es {n1+n2} de 5" if p2==True else f"Hay un error, vuelve a revisar tu codigo. Tu nota hasta este momento es {n1+n2} de 5" )
################################################################################################################
## Prueba de funcionamiento
## Esta celda no debe ser modificada
################################################################################################################

¡¡¡Felicidades!!! Tu nota hasta este momento es 2 de 5



<font color="darkblue"> 
3) Crear una lista llamada **meses_mayuscula** donde convierta todos los elementos de la lista **meses** usando la función `upper()` por medio del código `[mes.upper() for mes in meses]`. Crear una variable llamada **filtro_bre** donde se aplique un filtro sobre la variable **meses_mayusculas** usando la función `filter()` y `lambda`, tal que se filtren los meses que contengan el término "BRE". Después crear una variable llamada **filtro_ju** donde se aplique un filtro sobre la variable **meses_mayusculas** usando la función `filter()` y `lambda`, tal que se filtren los meses que contengan el término "JU". Por último, crear una variable llamada **lista_bre_ju** donde se unan las listas **lista_bre** y **lista_ju**.
</font>  
    


In [ ]:
################################################################################################################
## Este espacio debe ser llenado por el estudiante.
################################################################################################################

meses=['enero','febrero','marzo','abril','mayo','junio','julio','agosto','septiembre','octubre','noviembre','diciembre']
meses_mayusculas=[mes.upper() for mes in meses]
print(meses_mayusculas)
"""
Sintaxis filter
nueva_secuencia = filter(función_de_filtro, secuencia)

"""
filtro_bre=filter(lambda x:"BRE" in x ,meses_mayusculas)
lista_bre=list(filtro_bre)
print("Filtro bre")
print(lista_bre)
print("=========")
filtro_ju=filter(lambda x:"JU" in x,meses_mayusculas)
lista_ju=list(filtro_ju)
print("Filtro ju")
print(lista_ju)
print("=========")
print("Filtro ju + bre")
lista_bre_ju=lista_bre+lista_ju

print(lista_bre_ju)



['ENERO', 'FEBRERO', 'MARZO', 'ABRIL', 'MAYO', 'JUNIO', 'JULIO', 'AGOSTO', 'SEPTIEMBRE', 'OCTUBRE', 'NOVIEMBRE', 'DICIEMBRE']
Filtro bre
['FEBRERO', 'SEPTIEMBRE', 'OCTUBRE', 'NOVIEMBRE', 'DICIEMBRE']
Filtro ju
['JUNIO', 'JULIO']
Filtro ju + bre
['FEBRERO', 'SEPTIEMBRE', 'OCTUBRE', 'NOVIEMBRE', 'DICIEMBRE', 'JUNIO', 'JULIO']


In [ ]:
################################################################################################################
## Prueba de funcionamiento
## Esta celda no debe ser modificada
################################################################################################################

p31=set([mes.isupper() for mes in meses_mayusculas])==set([True]*12)
p32=len(lista_bre)==5
p33=len(lista_ju)==2
p34=len(lista_bre_ju)==7

#p3=p31 & p32 & p33 & p34  #p33
p3=p31 & p32 & p33 & p34 
if p3==True:
  n3=1
else:
  n3=0

print(f"¡¡¡Felicidades!!! Tu nota hasta este momento es {n1+n2+n3} de 5" if p3==True else f"Hay un error, vuelve a revisar tu codigo. Tu nota hasta este momento es {n1+n2+n3} de 5" )
################################################################################################################
## Prueba de funcionamiento
## Esta celda no debe ser modificada
################################################################################################################

¡¡¡Felicidades!!! Tu nota hasta este momento es 3 de 5



<font color="darkblue"> 
4) Se leyó el dataset **movies** de una url. A partir de una transformaciones se obtuvieron dos listas llamadas **cabecera** y **lineas_sin_cabecera**. Así pues, se pide:
- Crear una lista vacia llamada **nombre_pelicula**, luego usar el metodo `append()` y el metodo `split()` para seleccionar el nombre de las peliculas, tal y como se muestra en el siguiente ejemplo

    
    nombre_pelicula=[]
    for l in lineas_sin_cabecera:
        nombre_pelicula.append(l.split(',')[0])
    
   
<font color="darkblue"> 

- Crear listas vacias llamadas **anno** y **genero** y repetir el procedimiento anterior, seleccionando las variables **year** y **genre** respectivamente del dataset.
- Crear una lista vacia llamada **recaudacion_mundial** y aplicar el siguiente codigo:
</font>  
    
    
    for l in lineas_sin_cabecera:
        valor=re.findall(r'\d{1,3}.\d{1,2}\S',l.split(',')[-2])
        recaudacion=''.join(valor)
        recaudacion_mundial.append(recaudacion)
    
    i=recaudacion_mundial.index('')
    recaudacion_mundial[i]='0'
    recaudacion_mundial=[float(x) for x in recaudacion_mundial]

In [ ]:
################################################################################################################
## Lectura de datos
## Esta celda no debe ser modificada
################################################################################################################

import urllib.request
import re

url = 'https://gist.githubusercontent.com/tiangechen/b68782efa49a16edaf07dc2cdaa855ea/raw/0c794a9717f18b094eabab2cd6a6b9a226903577/movies.csv'

with urllib.request.urlopen(url) as datos:
   lineas=str(datos.readlines()).split('\\n')

lineas_sin_cabecera=[]
for linea in lineas:
    if linea==lineas[0]:
        cabecera=lineas[0][5:]
    else:
        if linea==lineas[len(lineas)-1]:
            lineas_sin_cabecera.append(linea[5:-1])
        else:
            lineas_sin_cabecera.append(linea[5:])
            
print(cabecera)
for l in lineas_sin_cabecera[:5]:
    print(l)
################################################################################################################
## Lectura de datos
## Esta celda no debe ser modificada
################################################################################################################

lm,Genre,Lead Studio,Audience score %,Profitability,Rotten Tomatoes %,Worldwide Gross,Year
Zack and Miri Make a Porno,Romance,The Weinstein Company,70,1.747541667,64,$41.94 ,2008
Youth in Revolt,Comedy,The Weinstein Company,52,1.09,68,$19.62 ,2010
You Will Meet a Tall Dark Stranger,Comedy,Independent,35,1.211818182,43,$26.66 ,2010
When in Rome,Comedy,Disney,44,0,15,$43.04 ,2010
What Happens in Vegas,Comedy,Fox,72,6.267647029,28,$219.37 ,2008


In [ ]:
################################################################################################################
## Este espacio debe ser llenado por el estudiante.
################################################################################################################
nombre_pelicula=[]
for l in lineas_sin_cabecera:
    nombre_pelicula.append(l.split(',')[0])
print(nombre_pelicula)
genero=[]
for l in  lineas_sin_cabecera:
    genero.append(l.split(',')[1])

print(genero)
anno=[]
for l in lineas_sin_cabecera:
   anno.append(l.split(',')[7])
print(anno)
recaudacion_mundial=None

recaudacion_mundial=[]
for l in lineas_sin_cabecera:
    valor=re.findall(r'\d{1,3}.\d{1,2}\S',l.split(',')[-2])
    recaudacion=''.join(valor)
    recaudacion_mundial.append(recaudacion)

i=recaudacion_mundial.index('')
recaudacion_mundial[i]='0'
recaudacion_mundial=[float(x) for x in recaudacion_mundial]


['Zack and Miri Make a Porno', 'Youth in Revolt', 'You Will Meet a Tall Dark Stranger', 'When in Rome', 'What Happens in Vegas', 'Water For Elephants', 'WALL-E', 'Waitress', 'Waiting For Forever', "Valentine's Day", "Tyler Perry's Why Did I get Married", 'Twilight: Breaking Dawn', 'Twilight', 'The Ugly Truth', 'The Twilight Saga: New Moon', "The Time Traveler's Wife", 'The Proposal', 'The Invention of Lying', 'The Heartbreak Kid', 'The Duchess', 'The Curious Case of Benjamin Button', 'The Back-up Plan', 'Tangled', 'Something Borrowed', "She's Out of My League", 'Sex and the City Two', 'Sex and the City 2', 'Sex and the City', 'Remember Me', 'Rachel Getting Married', 'Penelope', 'P.S. I Love You', 'Over Her Dead Body', 'Our Family Wedding', 'One Day', 'Not Easily Broken', 'No Reservations', "Nick and Norah's Infinite Playlist", "New Year's Eve", 'My Week with Marilyn', 'Music and Lyrics', 'Monte Carlo', 'Miss Pettigrew Lives for a Day', 'Midnight in Paris', 'Marley and Me', 'Mamma Mia!'

In [ ]:
################################################################################################################
## Prueba de funcionamiento
## Esta celda no debe ser modificada
################################################################################################################

import collections

p41=set(list(collections.Counter(genero).values()))==set([13, 41, 13, 4, 1, 1, 1, 1, 1, 1])
p42=set(list(collections.Counter(anno).values()))==set([20, 20, 14, 11, 11, 1])
p43=np.mean(recaudacion_mundial)==142.3733766233766
p44=len(nombre_pelicula)==77

p4=p41 & p42 & p43 & p44 
if p4==True:
  n4=1
else:
  n4=0

print(f"¡¡¡Felicidades!!! Tu nota hasta este momento es {n1+n2+n3+n4} de 5" if p4==True else f"Hay un error, vuelve a revisar tu codigo. Tu nota hasta este momento es {n1+n2+n3+n4} de 5" )

################################################################################################################
## Prueba de funcionamiento
## Esta celda no debe ser modificada
################################################################################################################

¡¡¡Felicidades!!! Tu nota hasta este momento es 4 de 5



<font color="darkblue"> 
5) Crear una lista llamada **tupla_atributos** donde cada componente de la lista sea una tupla que contenga los elementos de las variables **genero**, **anno** y **recaudacion_mundial**; por ejemplo
    
    
    
    [('Romance', '2008', 41.94),
    ('Comedy', '2010', 19.62),
    ('Comedy', '2010', 26.66),
    ('Comedy', '2010', 43.04),
    ('Comedy', '2008', 219.37),
    ('Drama', '2011', 117.09),
    ('Animation', '2008', 521.28),
    ('Romance', '2007', 22.18),...]

<font color="darkblue"> 
Crear una variable llamada **diccionario_peliculas** cuyas llaves sean los nombres de las peliculas y sus valores las tuplas de la lista **tupla_atributos**, ejemplo
    
    {'Zack and Miri Make a Porno': ('Romance', '2008', 41.94),
    'Youth in Revolt': ('Comedy', '2010', 19.62),
    'You Will Meet a Tall Dark Stranger': ('Comedy', '2010', 26.66),
    'When in Rome': ('Comedy', '2010', 43.04),
    'What Happens in Vegas': ('Comedy', '2008', 219.37),
    'Water For Elephants': ('Drama', '2011', 117.09),
    'WALL-E': ('Animation', '2008', 521.28),
    'Waitress': ('Romance', '2007', 22.18),
    'Waiting For Forever': ('Romance', '2011', 0.03),...}
    
Por último, crear un DataFrame, llamado **df**, usando la libreria pandas y usando el siguiente codigo
</font>  
    
    import pandas as pd
    nom_keys=list(diccionario_peliculas.keys())
    df=pd.DataFrame( [ diccionario_peliculas[key] for key in diccionario_peliculas.keys()],index= nom_keys,columns=[['genero','anno','recaudacion(millones_dolares)']]  )

In [ ]:
################################################################################################################
## Este espacio debe ser llenado por el estudiante.
################################################################################################################

diccionario_peliculas=None

tupla_atributos=[genero,anno,recaudacion_mundial]
tuplagen=tuple(tupla_atributos[0])
tupla_atributos=[tuplagen,anno,recaudacion_mundial]
tuplaanno=tuple(tupla_atributos[1])
tupla_atributos=[tuplagen,tuplaanno,recaudacion_mundial]
tuplamundial=tuple(tupla_atributos[2])
tupla_atributos1=[tuplagen,tuplaanno,tuplamundial]
tupla_atributos=np.array(tupla_atributos1); # estamos guardando el arreglo que genera numpy en una variable
num_filas = tupla_atributos.shape[0] # estamos guardando el numero de filas
num_columnas = tupla_atributos.shape[1] # estamos guardando el numero de columnas
print("el numero de filas es: ",num_filas)
print("el numero de columnas es: ",num_columnas)
print(tupla_atributos)
tupla_atributos =tupla_atributos.T
num_filas = tupla_atributos.shape[0] # estamos guardando el numero de filas
num_columnas = tupla_atributos.shape[1] # estamos guardando el numero de columnas
print("el numero de filas es: ",num_filas)
print("el numero de columnas es: ",num_columnas)
"""import pandas as pd
nom_keys=list(diccionario_peliculas.keys())
df=pd.DataFrame( [ diccionario_peliculas[key] for key in diccionario_peliculas.keys()],index= nom_keys,columns=[['genero','anno','recaudacion(millones_dolares)']]  )
"""
df=None


el numero de filas es:  3
el numero de columnas es:  77
[['Romance' 'Comedy' 'Comedy' 'Comedy' 'Comedy' 'Drama' 'Animation'
  'Romance' 'Romance' 'Comedy' 'Romance' 'Romance' 'Romance' 'Comedy'
  'Drama' 'Drama' 'Comedy' 'Comedy' 'Comedy' 'Drama' 'Fantasy' 'Comedy'
  'Animation' 'Romance' 'Comedy' 'Comedy' 'Comedy' 'Comedy' 'Drama'
  'Drama' 'Comedy' 'Romance' 'Comedy' 'Comedy' 'Romance' 'Drama' 'Comedy'
  'Comedy' 'Romance' 'Drama' 'Romance' 'Romance' 'Comedy' 'Romence'
  'Comedy' 'Comedy' 'Comedy' 'Comdy' 'Drama' 'Comedy' 'Comedy' 'Comedy'
  'Comedy' 'Comedy' 'Comedy' 'Action' 'Comedy' 'Romance' 'Comedy'
  'Comedy' 'Comedy' 'Comedy' 'Comedy' 'Comedy' 'Animation' 'Animation'
  'Comedy' 'Comedy' 'Drama' 'Comedy' 'Drama' 'Comedy' 'romance' 'Drama'
  'Drama' 'Comedy' 'comedy']
 ['2008' '2010' '2010' '2010' '2008' '2011' '2008' '2007' '2011' '2010'
  '2007' '2011' '2008' '2009' '2009' '2009' '2009' '2009' '2007' '2008'
  '2008' '2010' '2010' '2011' '2010' '2010' '2010' '2008' '2010' '2008

In [ ]:
                                                                                     ################################################################################################################
## Prueba de funcionamiento
## Esta celda no debe ser modificada
################################################################################################################

p51=np.array(tupla_atributos).shape==(77,3) and type(tupla_atributos)==list and type(tupla_atributos[0])==tuple # cambiando el orden de
p52=type(diccionario_peliculas)==dict and set(list(diccionario_peliculas.keys()))==set(nombre_pelicula) and set(list(diccionario_peliculas.values()))==set(tupla_atributos)
p53=type(df)==pd.core.frame.DataFrame and df.shape==(75,3) and set(list(df.columns))==set([('genero',), ('anno',), ('recaudacion(millones_dolares)',)])

p5=p51 &p52 &p53
if p5==True:
  n5=1
else:
  n5=0

print(f"¡¡¡Felicidades!!! Tu nota hasta este momento es {n1+n2+n3+n4+n5} de 5" if p5==True  else f"Hay un error, vuelve a revisar tu codigo. Tu nota hasta este momento es {n1+n2+n3+n4+n5} de 5" )

################################################################################################################
## Prueba de funcionamiento
## Esta celda no debe ser modificada
################################################################################################################

Hay un error, vuelve a revisar tu codigo. Tu nota hasta este momento es 4 de 5



# Comentarios por parte del estudiante
### Escriba aqui sus comentarios y conclusiones


-

-

-

